<a href="https://colab.research.google.com/github/aliffiazmuhammed/machine_learning_ict/blob/main/data_preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Data Pre-processing Stage

  This notebook contains the basic data pre processing steps.
  * Preprocessing refers to the transformations applied to the data before feeding it to the machine learning algorithms.
  * The data gathered from different sources is collected in raw format which is not feasible for the analysis.
  * Data Preprocessing technique is used to convert the raw data into a clean data set.

#### Why preprocessing ?
1. Real world data are generally
    * Incomplete: lacking attribute values, lacking certain attributes of interest, or containing only aggregate data.
    * Noisy: containing errors or outliers.
    * Inconsistent: containing discrepancies in codes or names.

Let's take a sample dataset for this exercise.
This dataset named "data.csv" contains whether a user purchased the product or not.
The users data has age,salary and the country they belonged to.

In [28]:
###############################################################
#       Step 1 : Importing the libraries                      #
###############################################################


# NumPy is module for Python. The name is an acronym for "Numeric Python" or "Numerical Python".
# This makes sure that the precompiled mathematical and numerical functions
# and functionalities of Numpy guarantee great execution speed.

import numpy as np

# Pandas is an open-source Python Library providing high-performance data manipulation
# and analysis tool using its powerful data structures.
# The name Pandas is derived from the word Panel Data – an Econometrics from Multidimensional data.

import pandas as pd


# The OS module in Python provides a way of using operating system dependent functionality.
# The functions that the OS module provides allows you to interface with the underlying operating system
# that Python is running on – be that Windows, Mac or Linux.
import sklearn
import os

In [8]:
pip install imputer


ERROR: Could not find a version that satisfies the requirement imputer (from versions: none)
ERROR: No matching distribution found for imputer


In [10]:
###############################################################
#       Step 2 : Importing the Dataset                        #
###############################################################

#Read the 'Data.csv' and store the data in the vairable dataset.
dataset = pd.read_csv("/content/preprocessing_data.csv")
print('Load the datasets...')


# Print the shape of the dataset
print ('dataset: %s'%(str(dataset.shape)))


Load the datasets...
dataset: (15, 4)


In [11]:
from google.colab import drive
drive.mount('/content/preprocessing_data.csv')

MessageError: Error: credential propagation was unsuccessful

The dataset contains 15 rows and 4 columns

In [12]:
# print the dataset
dataset

Country   Age   Salary Purchased
0       India  34.0  92000.0       Yes
1   Sri lanka  22.0  25000.0       Yes
2       China  31.0  74000.0       Yes
3   Sri lanka  29.0      NaN        No
4       China  55.0  98000.0       Yes
5       India  24.0  30000.0        No
6   Sri lanka  28.0  40000.0        No
7       India   NaN  60000.0        No
8       China  51.0  89000.0       Yes
9       India  44.0  78000.0       Yes
10  Sri lanka  21.0  20000.0        No
11      China  25.0  30000.0       Yes
12      India  33.0  45000.0       Yes
13      India  42.0  65000.0       Yes
14  Sri lanka  33.0  22000.0        No

In [22]:
# Separate the dependent and independent variables

# Independent variable
# iloc[rows,columns]
# Take all rows
# Take last but one column from the dataset (:-1)
X = dataset.iloc[:,:-1].values

# Dependent variable
# iloc[rows,columns]
# Take all rows
# Take last column from the dataset (:-1)
Y = dataset.iloc[:,3].values

In [23]:
# Print the X and Y
print ('X: %s'%(str(X)))
print ('-----------------------------------')
print ('Y: %s'%(str(Y)))

X: [['India' 34.0 92000.0]
 ['Sri lanka' 22.0 25000.0]
 ['China' 31.0 74000.0]
 ['Sri lanka' 29.0 nan]
 ['China' 55.0 98000.0]
 ['India' 24.0 30000.0]
 ['Sri lanka' 28.0 40000.0]
 ['India' nan 60000.0]
 ['China' 51.0 89000.0]
 ['India' 44.0 78000.0]
 ['Sri lanka' 21.0 20000.0]
 ['China' 25.0 30000.0]
 ['India' 33.0 45000.0]
 ['India' 42.0 65000.0]
 ['Sri lanka' 33.0 22000.0]]
-----------------------------------
Y: ['Yes' 'Yes' 'Yes' 'No' 'Yes' 'No' 'No' 'No' 'Yes' 'Yes' 'No' 'Yes' 'Yes'
 'Yes' 'No']


#### 1. Handle Missing Data

There are few missing data in the Age and salary columns (NaN values).

#### i. Deleting Rows:
*      We cannot remove the rows with the missing data as it will affect the output of the  machine learning algorithm.
*      However we can delete a particular row if it has a null value for a particular feature and a particular column if it has more than 70-75% of missing values.
      

#### ii. Replacing With Mean/Median/Mode:
*      This strategy can be applied on a feature which has numeric data like the age of a person.
*      We can calculate the mean, median or mode of the feature and replace it with the missing values.    
*     The loss of the data can be negated by this method which yields better results compared to removal of rows and  
*       columns.
*      Replacing with the above three approximations are a statistical approach of handling the missing values.
*     This method is also called as leaking the data while training.
*     Another way is to approximate it with the deviation of neighbouring values.
*     This works better if the data is linear.


In [24]:
###############################################################
#       Step 3 : Missing Data                                 #
###############################################################

# Scikit-learn provides a range of supervised and unsupervised learning algorithms via a consistent interface in Python.
# The sklearn.preprocessing package provides several common utility functions and transformer classes
# to change raw feature vectors into a representation that is more suitable for the downstream estimators.

import numpy as np
from sklearn.impute import SimpleImputer
# Imputer Class takes the follwing parameters:
#     missing_values : The missing values in our dataset are called as NaN (Not a number).Default is NaN
#     strategy       : replace the missing values by mean/median/mode. Default is mean.
#     axis           : if axis = 0, we take we of the column and if axis = 1, we take mean value of row.

imputer = SimpleImputer(missing_values=np.nan, strategy='mean')

# Fit the imputer on X.
# Take all rows and columns only with the missing values.
# Note: Index starts with 0. Upper bound (3) is not included.

# Fit imputer for columns 1 and 2 of X matrix.
imputer = imputer.fit(X[:,1:3])

#Replace missing data with mean of column
X[:,1:3] = imputer.transform(X[:,1:3])


In [25]:
print ('X: %s'%(str(X)))

X: [['India' 34.0 92000.0]
 ['Sri lanka' 22.0 25000.0]
 ['China' 31.0 74000.0]
 ['Sri lanka' 29.0 54857.142857142855]
 ['China' 55.0 98000.0]
 ['India' 24.0 30000.0]
 ['Sri lanka' 28.0 40000.0]
 ['India' 33.714285714285715 60000.0]
 ['China' 51.0 89000.0]
 ['India' 44.0 78000.0]
 ['Sri lanka' 21.0 20000.0]
 ['China' 25.0 30000.0]
 ['India' 33.0 45000.0]
 ['India' 42.0 65000.0]
 ['Sri lanka' 33.0 22000.0]]


* Mean Value of Age    = Sum of all age values /14   = 33.714285714285715
* Mean Value of Salary = Sum of all Salary value /14 = 54857.142857142855

#### 2. Encode the Categorical data

Categorical data are variables that contain label values rather than numeric values.
Some algorithms can work with categorical data directly.

For example, a decision tree can be learned directly from categorical data with no data transform required (this depends on the specific implementation).

Many machine learning algorithms cannot operate on label data directly. They require all input variables and output variables to be numeric.

This means that categorical data must be converted to a numerical form.

In our dataset there are 2 columns with categorical data.

The First column which contains the country and the last column purchased.

#### i.  Label Encoder:

    * It is used to transform non-numerical labels to numerical labels (or nominal categorical variables).
    * Numerical labels are always between 0 and n_classes-1.     

#### ii. OneHotEncoder:
    * Encode categorical integer features using a one-hot aka one-of-K scheme.
    * The input to this transformer should be a matrix of integers, denoting the values taken on by categorical (discrete)
      features.
    * The output will be a sparse matrix where each column corresponds to one possible value of one feature.
    * It is assumed that input features take on values in the range [0, n_values]
    * This encoding is needed for feeding categorical data to many scikit-learn estimators, notably linear models and SVMs
      with the standard kernels.        

In [29]:
###############################################################
#       Step 4 : Categorical variables                        #
###############################################################

from sklearn.preprocessing import LabelEncoder,OneHotEncoder

labelencoder_X = LabelEncoder()
X[:,0] = labelencoder_X.fit_transform(X[:,0])
X[:,0]

array([1, 2, 0, 2, 0, 1, 2, 1, 0, 1, 2, 0, 1, 1, 2], dtype=object)

Now the categorical data of the country value is changed to numerical value.

| Country | Value |
|:--------|:------|
| China   |   0   |  
| India   |   1   |   
| Srilanka|   2   |   


#### Dummy Encoding

    * The above encoding will result in a problem.
    * The label encoding transforms the data as shown in the table above.
    * The Machine learning algorithm will assume that China>India>Sri Lanka.
    * But this is not the case. We just converted the categorical value and assigned it to a numeric value.
    * Hence there is a need to apply Dummy encoding to the above dataset.

| Country | China | India | Sri Lanka |
|:--------|:------|:------|:----------|
| China   |   1   |  0    |    0      |   
| India   |   0   |  1    |    0      |   
| Srilanka|   0   |  0    |    1      |   
| India   |   0   |  1    |    0      |  
| Srilanka|   0   |  0    |    1      |  
| China   |   1   |  0    |    0      |  
  



In [37]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

# Assuming X is your dataset

# Create an instance of OneHotEncoder
encoder = OneHotEncoder()

# Define the columns to be one-hot encoded
# In this case, we specify [0] to indicate the first column
ct = ColumnTransformer(
    [('one_hot_encoder', encoder, [0])],  # The transformer to be applied
    remainder='passthrough'               # Leave the rest of the columns unchanged
)

# Apply the ColumnTransformer to your dataset X
X = ct.fit_transform(X)
X

array([[0.0, 1.0, 0.0, 1.0, 0.0, 34.0, 92000.0],
       [0.0, 1.0, 0.0, 0.0, 1.0, 22.0, 25000.0],
       [1.0, 0.0, 1.0, 0.0, 0.0, 31.0, 74000.0],
       [0.0, 1.0, 0.0, 0.0, 1.0, 29.0, 54857.142857142855],
       [1.0, 0.0, 1.0, 0.0, 0.0, 55.0, 98000.0],
       [0.0, 1.0, 0.0, 1.0, 0.0, 24.0, 30000.0],
       [0.0, 1.0, 0.0, 0.0, 1.0, 28.0, 40000.0],
       [0.0, 1.0, 0.0, 1.0, 0.0, 33.714285714285715, 60000.0],
       [1.0, 0.0, 1.0, 0.0, 0.0, 51.0, 89000.0],
       [0.0, 1.0, 0.0, 1.0, 0.0, 44.0, 78000.0],
       [0.0, 1.0, 0.0, 0.0, 1.0, 21.0, 20000.0],
       [1.0, 0.0, 1.0, 0.0, 0.0, 25.0, 30000.0],
       [0.0, 1.0, 0.0, 1.0, 0.0, 33.0, 45000.0],
       [0.0, 1.0, 0.0, 1.0, 0.0, 42.0, 65000.0],
       [0.0, 1.0, 0.0, 0.0, 1.0, 33.0, 22000.0]], dtype=object)

In [38]:
# Encoding the categorical data for Y matrix
labelencoder_Y = LabelEncoder()
Y = labelencoder_X.fit_transform(Y)
Y

array([1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0])

In [39]:
###############################################################
#       Step 5 : Splitting the dataset                        #
###############################################################

from sklearn.model_selection import train_test_split

# Assuming X and Y are your feature and target variables, respectively

# Split the dataset into training and testing sets
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=0)

In [40]:
# Print the shape of the dataset
print ('X_train: %s'%(str(X_train.shape)))
print ('----------------')
print ('X_test: %s'%(str(X_test.shape)))
print ('----------------')
print ('Y_train: %s'%(str(Y_train.shape)))
print ('----------------')
print ('Y_test: %s'%(str(Y_test.shape)))
print ('----------------')

X_train: (12, 7)
----------------
X_test: (3, 7)
----------------
Y_train: (12,)
----------------
Y_test: (3,)
----------------


#### 3. Scale your Features
    *  Most of the times, the dataset will contain features highly varying in magnitudes, units and range.
    *  Since the machine learning algorithms use Eucledian distance between two data points in their computations, this is
       result in wrong prediction.
      
We need to put the variables in same range, in the same scale so that no variable dominates the other variable.     
      
      

In [ ]:
###############################################################
#       Step 6 : Feature Scaling                              #
###############################################################

from sklearn.preprocessing import StandardScaler

sc_X = StandardScaler()

# We need to fit and transform the training set
X_train = sc_X.fit_transform(X_train)

# We need to fit the test set
X_test = sc_X.transform(X_test)

In [ ]:
X_train

In [ ]:
X_test

Now the all the data are in same scale. We can now apply different Machine learning model to the dataset.